engine/verification/fetcher/engine.go

```diff
@@ -94,9 +94,6 @@ func (e *Engine) WithChunkConsumerNotifier(notifier module.ProcessingNotifier) {
 
 // Ready initializes the engine and returns a channel that is closed when the initialization is done
 func (e *Engine) Ready() <-chan struct{} {
-	if e.chunkConsumerNotifier == nil {
-		e.log.Fatal().Msg("missing chunk consumer notifier callback in verification fetcher engine")
-	}
```

Comment:
The reason for this change is to resolve a cyclic dependency that causes node crash on startup: 

- In `cmd/verification/main`, the fetcher engine is both _created and started_ prior to the `chunkConsumer` (which has the `chunkConsumerNotifier`). 
- When `chunkConsumer` is created, it sets the `chunkConsumerNotifier` of `fetcher` engine. 
- So starting `fetcher` engine before `chunkConsumer` initialization crashes the node with a `fatal` error. 

Hence, we discard this checking and rely on the fact that _similar to any other component of a node_, missing a `chunkConsumerNotifier` literally leads the node to crash with panic upon processing the first chunk. This is the same strategy we take about other components, e.g., network, which missing assignment of that on a node panics the first time node tries sending a message. 

The other solution would be to decouple the component initialization and startup on the scaffold, but not sure of its side-effects on other nodes (considerable on a separate PR).

Reply:
This seems like it will introduce a race condition at startup where sometimes the node will crash. I don't think the network component has the same behaviour (it may error before the mesh network is set up but it won't crash the node). It would be better to avoid this if we can. 

What we do for some other engines with a similar dependency structure (eg. [`epochmgr`](https://github.com/onflow/flow-go/blob/663eaefd51d241f6ecf73415d3e1cf3f740527be/engine/collection/epochmgr/engine.go#L112-L136), [`compliance`](https://github.com/onflow/flow-go/blob/5635007f119cc67382fedae477452b0c00e27c24/engine/consensus/compliance/engine.go#L114-L142)) is embed the dependency within the engine and have the engine manage the lifecycle of the dependency. Then both the dependency and the engine can be instantiated in one `Component` block in the scaffold.

Reply:
I applied your comment in https://github.com/onflow/flow-go/pull/728/commits/955928152f74b7522feeb105be0bd17916dc2b2d, but realized the integration tests are failing due to a `panic` on verification startup:

```
      goroutine 213 [running]:
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       �github.com/onflow/flow-go/network/p2p.(*Network).Register(0x0, 0x156ba46, 0xe, 0x1823ae0, 0xc0001501e0, 0xc000388738, 0x9066a0, 0xc0000a5540, 0x1)
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       '	/app/network/p2p/network.go:118 +0x60
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       agithub.com/onflow/flow-go/engine/verification/requester.(*Engine).Ready(0xc0001501e0, 0x16eee10)
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       ;	/app/engine/verification/requester/requester.go:138 +0x72
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       Ngithub.com/onflow/flow-go/engine/verification/fetcher.(*Engine).Ready.func1()
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       6	/app/engine/verification/fetcher/engine.go:101 +0x39
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       [github.com/onflow/flow-go/engine.(*Unit).Ready.func1(0xc000238468, 0x1, 0x1, 0xc0003c6000)
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       	/app/engine/unit.go:106 +0x35
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       :created by github.com/onflow/flow-go/engine.(*Unit).Ready
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       	/app/engine/unit.go:104 +0x7f
··· DOCK: (setup ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) - container health failure: container not running
```

**Reason for panic**: 
As an engine, `requester` receives an instance of `network` and tries registering itself upon initialization. However, scaffold treats `network` as a `component` (not a module), and by default _all components are initialized and started post the initialization and startup of modules_. This means that when we treat `requester` engine as a `component` (instead of its original `module` nature), we are _enforcing_ its initialization _prior_ to all `component`s including `network` itself. So, by the time we are initializing `requester` engine as a `module` we pass a **network interface type with nil value** to it, which causes panic when it tries registering itself in the network. 

The log snippet below shows that `requester` engine, as a `module` is initialized _prior_ to the `network` layer as a `component`: 
```
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       �{"level":"info","node_role":"verification","node_id":"3c7a081647003c05e7dd4b2ca8144a4fec89b3fabf17d0ae5b55247d52ef1c2c","module":"sync core","time":"2021-05-20T20:50:32Z","message":"module initialization complete"}
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       �{"level":"info","node_role":"verification","node_id":"3c7a081647003c05e7dd4b2ca8144a4fec89b3fabf17d0ae5b55247d52ef1c2c","module":"requester engine","time":"2021-05-20T20:50:32Z","message":"module initialization complete"}
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       �{"level":"info","node_role":"verification","node_id":"3c7a081647003c05e7dd4b2ca8144a4fec89b3fabf17d0ae5b55247d52ef1c2c","module":"fetcher engine","time":"2021-05-20T20:50:32Z","message":"module initialization complete"}
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       �{"level":"info","node_role":"verification","node_id":"3c7a081647003c05e7dd4b2ca8144a4fec89b3fabf17d0ae5b55247d52ef1c2c","component":"network","time":"2021-05-20T20:50:32Z","message":"component initialization complete"}
··· DOCK: (clogs ) verification_1            (babf53b7e89f39b51543d3bac9398ae2d28f9c785e00b6de0242809b5615c650) -       �{"level":"info","node_role":"verification","node_id":"3c7a081647003c05e7dd4b2ca8144a4fec89b3fabf17d0ae5b55247d52ef1c2c","component":"network","time":"2021-05-20T20:50:32Z","message":"component startup complete"}
```

**Moving logic to startup instead of initialization**:
As a solution to the above panic case, one may suggest keeping the network as a member field in `requester` engine, let it receive a pointer to the network module, which will be started and initialized later. Then move the logic of network registration into `Ready` method of `requester` engine, so that we make sure we read the network value after its initialization is done. This solution is also results in the same panic (tried in https://github.com/onflow/flow-go/pull/728/commits/955928152f74b7522feeb105be0bd17916dc2b2d), and the reason is `module.Network` is an _interface type_, which initially has an interface value of `nil`. Now, when we pass a `nil`-value interface type to `requester.New`, we just are copying the `nil` into its member field interface type of `requester.net`, which is of the same type as `module.Network`. So, we are not copying an `nil` _pointer_ type that later on is initialized to the _address_ of a network object. Rather, we copy a `nil` interface type into another `nil` interface type: 

```
package requester 

func New(
        ....
	net module.Network, // a nil interface value at the time we initialize requester as module
        ....) (*Engine, error) {

	e := &Engine{
                ......
		net:              net, // copying a nil interface into another nil interface 
	}

	return e, nil
}


// Ready initializes the engine and returns a channel that is closed when the initialization is done.
func (e *Engine) Ready() <-chan struct{} {
        .....
        
	con, err := e.net.Register(engine.RequestChunks, e) // panics of nil e.net
	 
        ......

```

Reply:
Fixed in https://github.com/onflow/flow-go/pull/728/commits/955928152f74b7522feeb105be0bd17916dc2b2d and https://github.com/onflow/flow-go/pull/728/commits/892598a7a3d7f63587aefe82108b9b7c2cea4a40